File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.js
/**
 * @param {number[][]} points
 * @param {number} k
 * @return {number[][]}
 */
var kClosest = function(points, k) {
  let maxHeap = new MaxPriorityQueue({priority: (elem) => elem[0]});
  
  for (let i = 0; i < k; i++) {
    let [x, y] = points[i];
    let dist = (x * x) + (y * y)
    maxHeap.enqueue([dist, [x, y]]);
  }
  
  for (let i = k; i < points.length; i++) {
    let [x, y] = points[i];
    let dist = (x * x) + (y * y);
    
    if (dist < maxHeap.front().element[0]) {
      maxHeap.dequeue();
      maxHeap.enqueue([dist, [x, y]]);
    }
  }
  
  let res = [];
    
  for (let elem of maxHeap.toArray()) {
    res.push(elem.element[1])
  }
  
  return res;
};

File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: List[List[int]], K: int) -> List[List[int]]:
        maxH = []
        for i in range(len(points)):
            value = -(points[i][0]**2 + points[i][1]**2)
            heapq.heappush(maxH, (value, (points[i][0], points[i][1])))
            if len(maxH) > K:
                heapq.heappop(maxH)
        res = []
        while len(maxH)>0:
            res.append(maxH[0][1])
            heapq.heappop(maxH)
        return res
